components/dashboard/src/insights/download/download-sessions.ts

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright (c) 2024 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import {
+    ListWorkspaceSessionsRequest,
+    WorkspaceSession,
+    WorkspaceSpec_WorkspaceType,
+} from "@gitpod/public-api/lib/gitpod/v1/workspace_pb";
+import { organizationClient, workspaceClient } from "../../service/public-api";
+import dayjs from "dayjs";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { useCallback } from "react";
+import { noPersistence } from "../../data/setup";
+import { Timestamp } from "@bufbuild/protobuf";
+import type { OrganizationMember } from "@gitpod/public-api/lib/gitpod/v1/organization_pb";
+
+const pageSize = 100;
+const maxPages = 100; // safety limit if something goes wrong with pagination
+
+type GetAllWorkspaceSessionsArgs = Pick<ListWorkspaceSessionsRequest, "to" | "from" | "organizationId"> & {
+    signal?: AbortSignal;
+    onProgress?: (percentage: number) => void;
+};
+export const getAllWorkspaceSessions = async ({
+    from,
+    to,
+    signal,
+    organizationId,
+    onProgress,
+}: GetAllWorkspaceSessionsArgs): Promise<WorkspaceSession[]> => {
+    const records: WorkspaceSession[] = [];
+    let page = 0;
+    while (!signal?.aborted && page < maxPages) {
+        const response = await workspaceClient.listWorkspaceSessions(
+            {
+                organizationId,
+                from,
+                to,
+                pagination: {
+                    page,
+                    pageSize,
+                },
+            },
+            {
+                signal,
+            },
+        );
+        if (response.workspaceSessions.length === 0) {
+            break;
+        }
+
+        records.push(...response.workspaceSessions);
+        onProgress && onProgress(page);
+
+        page = page + 1;
+    }
+
+    return records;
+};
+
+type Args = Pick<ListWorkspaceSessionsRequest, "organizationId" | "from" | "to"> & {
+    organizationName: string;
+    signal?: AbortSignal;
+    onProgress?: (percentage: number) => void;
```

Comment:
Pagination is another issue that I faced here - `WorkspaceService.ListWorkspaceSessions` does not properly report a `total`, so we can't know how far in the process we are. For now, we just report "Exporting page ${n}".